### PR TITLE
Vueify Tool Form Wrapper

### PR DESCRIPTION
**Pull request by guerler**
Requires #11898 and is required to continue with #12227.

## What did you do? 
Currently the default Tool form wrapper is written in Backbone. Similar to #11898, this PR replaces the Tool form wrapper with Vue components which do not rely on the Backbone modules `tool-form-base` and `tool-form`. `tool-form-base` is currently inheriting `form-view`. The `form-view` module contains the Backbone recursion rendering methodology of all plain Galaxy forms which are build from a standardized JSON object. This PR does not replace the form recursion module but the customized modules which were build around it.

## Why did you make this change?
Our overall goal is to replace all form customization modules with Vue components. This will allow us to target the form recursion module itself and vueify it. Replacing legacy Backbone components with Vue components is a major effort to modernize the code base and comes with many benefits, such as improved code maintainability and reliability.

## How to test the changes? 
Select a tool from the Galaxy tool panel and execute it.

## Noticeable changes:
Errors / Warnings are now highlighted instantly in the UI using the `info` color. The idea behind this modification is to improve guiding the user while filling out the tool form.

### CODE REVIEW

File: client/src/components/Tool/ToolForm.vue
Comment (mvdbeek) [inline client/src/components/Tool/ToolForm.vue]: That should probably use the config and user provider

File: client/src/components/Tool/services.js
Comment (mvdbeek) [inline client/src/components/Tool/services.js]: It would be better to include the current history id here, since the tool building is context specific. We want to move away from using `trans.history` on the backend, which will be used if you don't provide a history id

Maybe `getToolFormData` as the method name would be more specific ?

File: client/src/components/Tool/ToolForm.vue
Comment (mvdbeek) [inline client/src/components/Tool/ToolForm.vue]: Can you move that into a provider ?

File: client/src/components/Tool/ToolForm.vue
Comment (mvdbeek) [inline client/src/components/Tool/ToolForm.vue]: That should probably be integrated with the beta history panel. Do you have a suggestion here @Nerdinacan ?

File: client/src/components/Tool/ToolForm.vue
Comment (mvdbeek) [inline client/src/components/Tool/ToolForm.vue]: Not sure that's the right way to get the current history id, and it should probably be consistent with what the tool form had been built with ?

File: client/src/components/Tool/ToolForm.vue
Comment (mvdbeek) [inline client/src/components/Tool/ToolForm.vue:278]: Should these be events that are handled externally ?

File: client/src/components/Tool/ToolForm.vue
Comment (Nerdinacan) [reply client/src/components/Tool/ToolForm.vue]: Yes. Either separate out the CurrentUser and Config providers or use the corresponding vuex store methods directly. I think separation is better. Instead of one big component this should be one big component and one small component whose job is to assemble dependencies like this:

```html
<!-- instead of --->
<ToolForm v-bind="yourProps" />

<!-- New Top-level component separates depenency collection from form rendering -->
<template>
    <ConfigProvider v-slot="{ config }">
        <CurrentUser v-slot="{ user }">
            <!-- this is basically the existing ToolForm with all the dependencies removed and turned into props
            You can transparently bind events and properties with the magic keywords $attrs and $listeners as I've 
            shown below.  To the outside, this will look exactly like ToolForm looks now with all the props and 
            event handlers passing through to your original form  -->
            <PreviousToolForm :config="config" :user="user" v-bind="$attrs" v-on="$listeners" />
        </CurrentUser>
    </ConfigProvider>
</template>
```

File: client/src/components/Tool/services.js
Comment (Nerdinacan) [inline client/src/components/Tool/services.js]: Can you get whatever Galaxy.params are from the config instead of using Galaxy here?

File: client/src/components/Form/Elements/FormBoolean.test.js
Comment (Nerdinacan) [inline client/src/components/Form/Elements/FormBoolean.test.js]: Why use a custom event name? I see this throughout this PR. You are often using onChange instead of the standard v-bind and/or .sync syntax. This is adding extra required methods to keep track of and I'm not seeing any new functionality not provided by v-model or .sync. Is there a reason for that?

File: client/src/components/Form/Elements/FormBoolean.vue
Comment (Nerdinacan) [inline client/src/components/Form/Elements/FormBoolean.vue]: Why not $emit('input') so you can use v-model, or $emit('update:value') so you can use :value.sync="externalVal" ?

File: client/src/components/Form/Elements/FormInput.vue
Comment (Nerdinacan) [inline client/src/components/Form/Elements/FormInput.vue]: Same v-model/vsync situation

**Comment by Nerdinacan** [comment]
Please rename the component and/or module "Form" to something more specific that actually reflects its intended function. Maybe "ToolInputForm", "GalaxyToolForm", "ToolFormInputTree"? You choose, but make it something specific.

I know that you are cartying over the old nomenclature from backbone, but the old nomenclature is bad and this is your chance to fix that. This kind of vagueness is irritating to the programmers who come after you.

"Form" is also the same as existing native javascript object, which may indeed create collisions depending on the means by which this script ultimately hits the page. 

The same policy should go for Element, Input, Text, etc. Make them all specific to the functionality which you are designing.

File: client/src/components/Form/Elements/FormBoolean.test.js
Comment (Nerdinacan) [inline client/src/components/Form/Elements/FormBoolean.test.js]: Are you using app?

File: client/src/components/Form/Form.vue
Comment (Nerdinacan) [inline client/src/components/Form/Form.vue]: Same v-model issue

File: client/src/components/Form/FormCard.test.js
Comment (Nerdinacan) [inline client/src/components/Form/FormCard.test.js]: Why?

File: client/src/components/Form/FormCard.vue
Comment (Nerdinacan) [inline client/src/components/Form/FormCard.vue:29]: Add a default if your prop is not required

File: client/src/components/Form/FormElement.vue
Comment (Nerdinacan) [inline client/src/components/Form/FormElement.vue]: See v-model vs. onChange comments

File: client/src/components/Panels/ToolBox.vue
Comment (Nerdinacan) [inline client/src/components/Panels/ToolBox.vue:35]: You might want to adopt a consistent event/handler naming convention. Typically the handler will be named "onSomething" and the event will be named "something"

File: client/src/components/Tool/ToolCard.test.js
Comment (Nerdinacan) [inline client/src/components/Tool/ToolCard.test.js]: Why?

File: client/src/components/Tool/ToolCard.vue
Comment (Nerdinacan) [inline client/src/components/Tool/ToolCard.vue]: If this uses the standard mount, all these labels can be localized with the "L" or "localize" filters or the v-localize directive.

```html
<b-dropdown-item @click="onCopyLink">
<span class="fa fa-chain" />
<span v-localize>Copy Link</spam>
</b-dropdown-item>
```

File: client/src/components/Tool/ToolForm.vue
Comment (Nerdinacan) [inline client/src/components/Tool/ToolForm.vue]: v-model?

File: client/src/components/Tool/ToolForm.vue
Comment (Nerdinacan) [reply client/src/components/Tool/ToolForm.vue]: I made a comment in the main thread, but my suggestion is to take ToolForm and separate out its global dependencies in a wrapper component then put ToolForm inside. But yes, currentHistory, config, user, and any other one-time lookup source data for this component should come either from Vuex directly or (more safely) from the providers. The providers ensure the values you want will be initialized before you try to use them.

File: client/src/components/Tool/ToolForm.vue
Comment (Nerdinacan) [inline client/src/components/Tool/ToolForm.vue]: CurrentUser provider? Have a user property for this component.

File: client/src/components/Tool/ToolForm.vue
Comment (Nerdinacan) [reply client/src/components/Tool/ToolForm.vue]: The real problem is that it is in a method which must be manually called. The value will not be dynamic as the currentHistory changes from outside this component. History should probably be a property of the component and its value should be pulled from Vuex, or a provider.

More modularaly, when the ToolForm is done, the component could emit a "complete" event with the payload and whoever is using the ToolForm could decide how to save it and where to get its history value. I think the whole ToolForm should act like a big v-model whose emitted value is a nested object of inputs, then whoever is using the ToolForm could decide how to use that input object, probably with the save function that requires the history.

File: client/src/components/Tool/ToolForm.vue
Comment (Nerdinacan) [reply client/src/components/Tool/ToolForm.vue:278]: @mvdbeek  I would argue yes. We should be emitting events and the outside should decide how to deal with them.

File: client/src/components/Tool/utilities.js
Comment (Nerdinacan) [inline client/src/components/Tool/utilities.js]: This send function is the only reason jQuery is imported and I'm not clear on why this function has to exist in the first place.

Here's something similar from Upload with just axios and native Form objects.

```js

import axios from "axios";
import { prependPath } from "utils/redirect";
import { UploadItem } from "../model";
import { History } from "components/History/model";

const api = axios.create({
    baseURL: prependPath("/api"),
});

export async function saveItemToHistory(opts = {}) {

    // ...bunch of other stuff omitted

    let response;
    try {
        const body = new FormData();
        body.set("tool_id", "upload1");
        body.set("history_id", history.id);

        response = await api.post("/tools", body);
    } catch (err) {
        throw new SaveItemError(item, history, err.message);
    }
    return response;
}


```

**Review by Nerdinacan — CHANGES_REQUESTED** [review]

File: client/src/components/Tool/ToolForm.vue
Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: Not sure what you mean with "save" function but overall we are on the same page regarding the goal towards we are moving. Is there a history provider I can use, or should I add a preliminary one within this PR?

File: client/src/components/Tool/ToolForm.vue
Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: Added config and user providers.

File: client/src/components/Tool/services.js
Comment (guerler) [reply client/src/components/Tool/services.js]: Added the history_id to the service helper and renamed it as suggested. Requires history provider. See discussion below.

File: client/src/components/Tool/ToolForm.vue
Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: I think thats a great idea.

File: client/src/components/Form/FormCard.test.js
Comment (guerler) [reply client/src/components/Form/FormCard.test.js]: Thanks for catching it, this is a left over from a test example elsewhere.

File: client/src/components/Form/FormCard.vue
Comment (guerler) [reply client/src/components/Form/FormCard.vue:29]: Good catch, thanks.

File: client/src/components/Tool/ToolCard.vue
Comment (guerler) [reply client/src/components/Tool/ToolCard.vue]: Done.

File: client/src/components/Tool/ToolForm.vue
Comment (mvdbeek) [reply client/src/components/Tool/ToolForm.vue]: There is the UserHistories provider, but maybe the watches in there aren't right for this situation. I guess a separate CurrentHistoryId provider might make sense ? It can take the value from the betaHistory vuex store (https://github.com/galaxyproject/galaxy/blob/11363a013ee187baaa0e7733ed909dfa791f38be/client/src/components/History/model/historyStore.js#L46)

File: client/src/components/Tool/services.js
Comment (guerler) [reply client/src/components/Tool/services.js]: Galaxy.params contains url parameters parsed through the backend. This is legacy code which we probably do not need anymore. Instead we can use `URLSearchParams` to access the parameters directly from the client.

File: client/src/components/Tool/ToolForm.vue
Comment (OlegZharkov) [inline client/src/components/Tool/ToolForm.vue]: It appears to be not a function now

```suggestion
            return config.server_mail_configured && !user.isAnonymous;
```

Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: Thanks @OlegZharkov

File: client/src/components/Tool/ToolForm.vue
Comment (Nerdinacan) [reply client/src/components/Tool/ToolForm.vue]: @mvdbeek 
 I see two watches in UserHistories, one which reloads the list of histories if the user changes and one which reloads the history data if the user selects a new history. Is one or the other of these not behaving well?

@guerler 
I was just suggesting that if the form emitted a "complete" or even just the v-model "input" event when the data was deemed ready, then you could modularly separate the components whose job is to get the big object of data (your form components) and the code that decided what to do with that data. I was using "save" as a generic term to indicate "whatever you do with the form data". 

Here's a bunch of sample use cases for your form all of which require an event to be emitted to separate the concerns of data collection from the concerns of "whatever the upstream component wants to do with that data".

```vue
<template>
    <!-- 
    Here I've broken v-model into value/@input to show how you could decide to use the returning
    data in any number of ways if the end result were presented as a Vue event, even if just as the built-in
    event name that comes baked-in with v-model.  With components, it is the use of events that makes
    separation of concerns possible. It will make your components more usable, testable, and flexible. -->

    Option 1: Give some inputs and run a function when the user edits those inputs
    <ToolForm :config="formConfiguration" :value="inputs" @input="thisIsTheSaveFunction" />

    Option 2: bind toolForm data to local temp var
    <ToolForm :config="someOtherConfig" v-model="bigObjectStoredLocally" />

    Option 3: Use the ToolForm as just one part of a larger form and forward its results to yet another outside component
    <input v-model="someNonConfiguredVal" />
    <ToolForm :config="subTree" v-model="configuredInputs" />
</template>

<script>
export default {
    data() {
        return { 
            // Option 3: part of a bigger form
            someNonConfiguredVal: "foo",
            configuredInputs: {},

            // Option 2: binds toolForm data to local temp var
            // now you've got a "toolForm" whose job is just to edit whatever this temp value is for
            // you could watch it and perform an update, or just emit some calculated value from
            // the return result, You get to decide, but decide OUT HERE, and not inside ToolForm
            bigObjectStoredLocaly: {}
        },
    },
    computed: {
        // Option 3
        compositeFormData() {
            return { ...this.configuredInputs, someNonConfiguredVal: this.someNonConfiguredVal };
        }
    },
    watch: {
        // Option 3: composite form result
        compositeFormData(newResult) {
              this.$emit('input', newResult);
        }
    },
    methods: {
        // Option 1: respond to changes in your editor
        thisIsTheSaveFunction(valFromForm) {
             // You could save the value to the server, start a tool running, update something in the store,
             // whatever you need to do. The point is to decouple the data retrieval from whatever you do 
             // with the data. This function  lives out here, and is not hard-baked inside the Form component
             // This is how you separate concerns with components, you just put the concerns in different components.
         }
    }
}
</script>
```

File: client/src/components/Tool/ToolForm.vue
Comment (OlegZharkov) [reply client/src/components/Tool/ToolForm.vue]: @guerler can you take a look in your gitter please? :)

File: client/src/components/Tool/utilities.js
Comment (guerler) [reply client/src/components/Tool/utilities.js]: @OlegZharkov pointed this code path out. After looking into the history of this, I noticed that this is a legacy code path which allows tools to submit the resulting form data to an arbitrary url. Since `data source` tools do not display the tool form, this path seems to be unnecessary, hence I removed it.

**Comment by guerler** [comment]
@Nerdinacan @mvdbeek Thanks for the comments. I think we addressed most of them here and will continue to address the remaining ones in follow-up PRs. Some require the new history to be the default history and others require the analysis router to be able to listen to events, ideally once the backbone router has been converted to Vue.

File: client/src/components/Tool/ToolForm.vue
Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: I embedded the history provider but the history refresh should remain as is until the beta history is activated by default.

File: client/src/components/Tool/ToolForm.vue
Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: This is pretty much how it looks. In the last stage, we can revisit all locations in which the `FormDisplay` component is used and we can use `input` event instead of the `change` event and further smooth out the component interface.

File: client/src/components/Form/FormElement.vue
Comment (guerler) [inline client/src/components/Form/FormElement.vue:55]: Just seeing that this change here, causes a dependency between the element wrapper and the actual type. This needs to be resolved before merging.

File: client/src/components/Tool/ToolForm.vue
Comment (mvdbeek) [reply client/src/components/Tool/ToolForm.vue]: Can you explain why ? You're passing in the history id, which should be synched, regardless of whether you're using the old or new history panel.

Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: Right, the `history id` comes from the new provider and works fine, but we are currently making additional manual calls to refresh the history through the `Galaxy` object. We also listen directly to changes in the current history, both of these functions have been transferred from the legacy code. Once the new history is switched on by default we can remove these insertions from the `ToolForm` component.

File: client/src/components/Tool/ToolForm.vue
Comment (guerler) [reply client/src/components/Tool/ToolForm.vue:278]: Yes, please see my comment above. Also currently the `ToolForm` component is directly loaded from the `Analysis` router which has not been converted to Vue yet. These two insertions which rely on the global Galaxy object are only required while the current history panel is on by default.

File: client/src/components/Tool/ToolForm.vue
Comment (mvdbeek) [reply client/src/components/Tool/ToolForm.vue]: But the history provider should always have to correct current history id, why do we need to listen to history panel twice ? Or does that not work ? What manual manipulations are we talking about, I had the impression that this was completely synced now.

Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: 1) After submitting a job, we trigger a refresh of the history panel. This is required to quickly show the new datasets in the history panel.
2) When a dataset is added to the legacy history, the tool form needs to request an update in order to display the new datasets instantly in the data selector.

Comment (mvdbeek) [reply client/src/components/Tool/ToolForm.vue]: Alright, but don't we need to continue doing that when we switch to the beta history ? Sounds like the tool form would still need to be rebuilt, and we'd still like to reset any history refresh backoff ? IOW it might be a good idea to think about this now ?

Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: Once the new history is active, we can listen to its state change by listening to regular events emitted from the corresponding component or provider. I think we should keep on moving forward with the transition, ideally early in the release cycle, once #12227 is merged too. All three top-level form wrappers will be entirely vueified. At that point `tool-form-base` will be gone and it will be easy for us to align the interaction between the history and the tool form, and focus on the design of input elements (i.e. v-model, vs .sync etc, which we somewhat unfortunately already mixed up in here) and their interface with the `FormDisplay` component before replacing the form recursion itself.

Comment (mvdbeek) [reply client/src/components/Tool/ToolForm.vue]: We can move forward, but what you're describing are events that the history cannot know about, it will not emit events when a job is submitted, since it has no idea that was done. That's an event that either the form needs to emit and that the provider listens to (or the provider provides a refresh callback ?)

Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: I am describing the currently existing framework and the two events which made it through our vuefication process. I would imagine that once the job is submitted a web socket? tells the history to update. Otherwise we could just remove the update request and accept a short delay until the continuous history request triggers an update by itself, either way my emphasis is that this is not really related to this PR. The goal of these PRs is to safely transition and remove `tool-form-base`, not necessarily to find a better way to deal with the interaction between the `tool-form` and the old or new `history`.

**Review by mvdbeek — APPROVED** [review]

File: client/src/components/Tool/ToolForm.vue
Comment (mvdbeek) [reply client/src/components/Tool/ToolForm.vue]: I'm not sure it's a good idea to ignore these questions while modernizing a component, but I trust your plan, so let's go with it.

Comment (guerler) [reply client/src/components/Tool/ToolForm.vue]: Really not trying to ignore those two legacy events but as I said my suggestion is to move in concise steps. The tool-form/history interaction should be a separate PR, particularly considering that we will hopefully soon have a brand new history and that this only relates too 2-3 lines of code which already exist in the current code base.

**Comment by Nerdinacan** [comment]
@guerler Awesome, most of my comments were intended to be suggestions rather than found errors. I hope they were helpful!

**Comment by dannon** [comment]
@guerler I'll take a quick look through this and get it merged today.  Thanks for the reviews Marius and Mason!

**Pull request merged**
